core/src/main/java/io/questdb/std/DirectIntList.java

```diff
@@ -0,0 +1,185 @@
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.std;
+
+import io.questdb.cairo.Reopenable;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.str.CharSink;
+
+import java.io.Closeable;
+
+public class DirectIntList implements Mutable, Closeable, Reopenable {
+
+    private static final Log LOG = LogFactory.getLog(DirectIntList.class);
+    private final long initialCapacity;
+    private final int memoryTag;
+    private long address;
+    private long capacity;
+    private long limit;
+    private long pos;
+    private long start;
+
+    public DirectIntList(long capacity, int memoryTag) {
+        this.memoryTag = memoryTag;
+        this.capacity = (capacity * Integer.BYTES);
+        this.address = Unsafe.malloc(this.capacity, memoryTag);
+        this.start = this.pos = address;
+        this.limit = pos + this.capacity;
+        this.initialCapacity = this.capacity;
+    }
+
+    public void add(int x) {
+        ensureCapacity();
+        assert pos < limit;
+        Unsafe.getUnsafe().putInt(pos, x);
+        pos += Integer.BYTES;
+    }
+
+    public final void add(DirectIntList that) {
+        long thatSize = that.pos - that.start;
+        if (limit - pos < thatSize) {
+            setCapacityBytes(this.capacity + thatSize - (limit - pos));
+        }
+        Vect.memcpy(this.pos, that.start, thatSize);
+        this.pos += thatSize;
+    }
+
+    // clear without "zeroing" memory
+    public void clear() {
+        pos = start;
+    }
+
+    public void clear(int b) {
+        zero(b);
+        pos = start;
+    }
+
+    @Override
+    public void close() {
+        if (address != 0) {
+            Unsafe.free(address, capacity, memoryTag);
+            address = 0;
+            start = 0;
+            limit = 0;
+            pos = 0;
+            capacity = 0;
+        }
+    }
+
+    public int get(long p) {
+        return Unsafe.getUnsafe().getInt(start + (p << 2));
+    }
+
+    // base address of native memory
+    public long getAddress() {
+        return address;
+    }
+
+    // capacity in INTs
+    public long getCapacity() {
+        return capacity / Integer.BYTES;
+    }
+
+    @Override
+    public void reopen() {
+        if (address == 0) {
+            resetCapacity();
+        }
+    }
+
+    public void resetCapacity() {
+        setCapacityBytes(initialCapacity);
+    }
+
+    public void set(long p, int v) {
+        assert p >= 0 && p <= (limit - start) >> 2;
+        Unsafe.getUnsafe().putInt(start + (p << 2), v);
+    }
+
+    // desired capacity in LONGs (not count of bytes)
+    public void setCapacity(long capacity) {
+        setCapacityBytes(capacity * Integer.BYTES);
+    }
+
+    public void setPos(long p) {
+        assert p * Integer.BYTES <= capacity;
+        pos = start + p * Integer.BYTES;
+    }
+
+    public void shrink(long newCapacity) {
```

Comment:
not tested

Reply:
also it's unclear whether the capacity is in bytes or in number of int slots

Reply:
Fixed a bug and added a test in 4f6970662d42b81a7edd80f348c73eb4ed238e34